### PR TITLE
[ROCM][CI] Introduce tests-to-include as rocm-test workflow input

### LINKED ISSUE #110181: CI: rocm `(default, 1, 3, linux.rocm.gpu)` is very slow
## Current Status
ongoing

## Issue looks like
Taking ~2.5 hours on https://github.com/pytorch/pytorch/pull/110167
image

Taking 3.5+ hours on https://github.com/pytorch/pytorch/pull/109976
image

## User impact
Slower merging

## Root cause
Seems like it may have been introduced in https://github.com/pytorch/pytorch/pull/109817 @malfet 
image

## Mitigation
Not sure

## Prevention/followups
Investigate cause of slow running time or split up tests into smaller test jobs. Try to make the tests run in similar time to CUDA tests (~1.5 hours)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @seemethere @malfet @pytorch/pytorch-dev-infra @ZainRizvi @kit1980 @huydhn @clee2000
```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by jithunnair-amd**
Fixes https://github.com/pytorch/pytorch/issues/110181

cc @jeffdaily @sunway513 @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Comment by jithunnair-amd** [comment]
Seeing a failure `2023-10-17T00:59:20.6930616Z FAILED [0.0036s] test_cuda.py::TestCudaMallocAsync::test_allocator_settings - RuntimeError: Unrecognized CachingAllocator option: release_lock_on_cudamalloc` that is probably already fixed upstream ... 
@pytorchbot rebase

**Comment by jithunnair-amd** [comment]
ROCm CI ran 1 shard as expected: https://github.com/pytorch/pytorch/actions/runs/6552121852/job/17795441333
Duration: 44mins
Python tests run: test_cuda (78s), test_torch (160s)
Cpp tests run: ~30m!

Since the aim of this PR is to introduce a way to run smaller set of core unit tests, we should exclude the cpp tests.

File: .ci/pytorch/test.sh [hunk 1]
```diff
@@ -228,13 +233,13 @@ test_python_shard() {
     exit 1
   fi
 
-  time python test/run_test.py --exclude-jit-executor --exclude-distributed-tests --shard "$1" "$NUM_TEST_SHARDS" --verbose
+  time python test/run_test.py --exclude-jit-executor --exclude-distributed-tests "$INCLUDE_CLAUSE" --shard "$1" "$NUM_TEST_SHARDS" --verbose
```
Comment (huydhn) [inline .ci/pytorch/test.sh:236]: Could you try to use `python test/run_test.py --exclude-jit-executor --exclude-distributed-tests --include $TESTS_TO_INCLUDE --verbose` here? As the correct syntax is `python test/run_test.py --exclude-jit-executor --exclude-distributed-tests --include test_torch test_cuda --verbose`, one `--include` and without quote.

The other option would be to update `run_test.py` itself to handle quoted list of tests I think.

Comment (jithunnair-amd) [reply .ci/pytorch/test.sh]: @huydnn I went with the shellcheck exception, and it passes lint now.

**Comment by jithunnair-amd** [comment]
@huydhn @clee2000 The ROCm CI as part of `ciflow/trunk` passed and ran the reduced set of tests in 53m: https://github.com/pytorch/pytorch/actions/runs/6590967469/job/17909206391

`test_nn test_torch test_cuda test_ops test_unary_ufuncs test_binary_ufuncs test_autograd`

**Review by huydhn — APPROVED** [review]

**Comment by huydhn** [comment]
Ping @jithunnair-amd to see if there is any update on this one

**Comment by jithunnair-amd** [comment]
> Ping @jithunnair-amd to see if there is any update on this one. The context is that we start to see some ROCm failures landing in trunk, i.e. https://hud.pytorch.org/pytorch/pytorch/commit/7ccca60927cdccde63d6a1d40480950f24e9877a, because the PR didn't have `ciflow/rocm`

Just updated this PR to use ROCm5.7, but otherwise it looks good from my end if all ROCm tests pass.

Requesting @jeffdaily to also take a look in case I'm missing something.

**Comment by jithunnair-amd** [comment]
> > @jithunnair-amd are the CI failures real?
> 
> The torchvision build failure is real, but is due to some unsupported compiler flags:
> 
> ```
> clang++: error: unsupported option '--generate-dependencies-with-compile'
> clang++: error: unsupported option '--dependency-output'
> ```
> 
> The previous CI run succeeded and used a different torchvision commit. There was another torchvision commit bump after the one in the most recent failing CI run. I'm assessing if that'll resolve this issue.

Actually, https://github.com/pytorch/pytorch/commit/0a7eef9bcf6d1f8b5531102342ffc21f24beb58d fixed the issue wrt unsupported compiler flags, so expecting a rebase to help.

**Comment by jithunnair-amd** [comment]
ROCm CI passing with rebase: https://github.com/pytorch/pytorch/actions/runs/6854687390/job/18638805996

Merging as pre-rebase commit had all CI checks passing except ROCm.

@pytorchbot merge -f "ROCm CI check passed post-rebase; all other CI checks passed pre-rebase already"

**Pull request closed without merging**